src/hotspot/share/utilities/filterQueue.inline.hpp

```diff
@@ -111,4 +111,29 @@ E FilterQueue<E>::pop(MATCH_FUNC& match_func) {
   } while (true);
 }
 
+// MT-Unsafe, external serialization needed.
+template <class E>
+template <typename MATCH_FUNC>
+E FilterQueue<E>::peek(MATCH_FUNC& match_func) {
+  Node*  first       = load_first();
+  Node*  cur         = first;
+  Node*  match       = NULL;
+
+  if (cur == NULL) {
+    return (E)NULL;
+  }
+  do {
+    if (match_func(cur->_data)) {
+      match = cur;
```

Comment:
Hmmm.... I was expecting a break after this match.
Is there a reason to continue the loop?

But now I'm realizing that I'm confused since this is called `peek()`, but
it's not really peeking at the head of the queue. It is searching the entire
queue for a match.

Reply:
It is a FIFO queue, which provides inserts at the end and retrieves at the front of the queue (push/pop/peek).
Push inserts at tail and pop/peek looks at head. (maybe these should be named differently to reflect this better, but that is another PR :) )

To provide lock-free inserts the queue only have a tail pointer.
This enabled us to do concurrent inserts with other inserts and/or another serialized operation, such as peek.
To get to the front of the queue we must always walk until the end.

This means peek is peeking at the head (with a filter).
Conceptually we generate a queue from that filter and from that virtual queue we peek at it's head.

Tail->A1->A2->A3->A4->A5
Filter = even
Virtual tail->A2->A4
Virtutal head = A4

To get to A4 we walk and apply the filter, the last match will be the head we are looking to peek at.

Reply:
So it's not just find _a_ match. It is find the _last_ match. OK.